Extractor/Connect/SessionManager.cs

```diff
@@ -0,0 +1,556 @@
+                                state = SessionManagerState.Connecting;
+                                continue;
+                            }
+
+                            try
+                            {
+                                newConnection = await HandleConnectResult(r, token);
+                            }
+                            catch (Exception ex)
+                            {
+                                ExtractorUtils.LogException(log, ex, "Unexpected error when handling connect result after service level change");
+                                await CloseSession(r.Connection.Session, token);
+                                state = SessionManagerState.Connecting;
+                                continue;
+                            }
+                        }
+                    }
+                    if (newConnection && !isInitial)
+                    {
+                        // Don't run the callbacks inside the session manager loop, since
+                        // if we lose connection while the callbacks are running,
+                        // we need the loop to recover.
+                        client.Callbacks.TaskScheduler.ScheduleTask(null, _ => client.Callbacks.OnServerReconnect(client));
+                    }
+
+                    await onNewPendingActions.WaitAsync(token);
+                }
+            }
+        }
+
+        public static async Task<byte> ReadServiceLevel(ISession session, CancellationToken token)
+        {
+            var res = await session.ReadAsync(null, 0, TimestampsToReturn.Neither, new ReadValueIdCollection
+            {
+                new ReadValueId
+                {
+                    NodeId = VariableIds.Server_ServiceLevel,
+                    AttributeId = Attributes.Value
+                }
+            }, token);
+            var dv = res.Results[0];
+            return dv.GetValue<byte>(0);
+        }
+
+        private async Task<bool> EnsureConnected(CancellationToken token)
+        {
+            if (state == SessionManagerState.Connected) return false;
+
+            var result = await connectionSource.Connect(connection, false, appConfig, token);
+
+            try
+            {
+                return await HandleConnectResult(result, token);
+            }
+            catch
+            {
+                await CloseSession(result.Connection.Session, token);
+                state = SessionManagerState.Connecting;
+                connection = null;
+                throw;
+            }
+        }
+
+        private async Task<bool> UpdateServiceLevel(byte newLevel, bool fromConnection)
+        {
+            var oldServiceLevel = CurrentServiceLevel;
+
+            if (newLevel != oldServiceLevel)
+            {
+                log.LogDebug("Server ServiceLevel updated {From} -> {To}", oldServiceLevel, newLevel);
+            }
```

Comment:
else return false? All the rest of the method seems to be dealing with level change.

Reply:
That's fair.